### ISSUE TITLE
Register a netCDF media type

### ISSUE THREAD

**Issue by ethanrd** [Unidata/netcdf#42]
Information on media types and the registration process is here: http://www.iana.org/assignments/media-types/.

I believe the next step is to draft some text for the registration proposal.

Some references to previous conversations on a netCDF media type are listed in opengeospatial/netcdf-ld#80. Thanks to @marqh for starting this conversation again.

**Comment by marqh** [comment]
Hello @ethanrd 

many thanks for raising this issue

some thoughts from me on elements within 
https://www.iana.org/form/media-types
(ref: https://www.rfc-editor.org/rfc/rfc6838.html )

marqh

* type name:
    * application
* subtype name
    * standards tree (no prefix)
* Required / optional Parameters
    * this could be left blank, as per rfc6838
* Encoding Considerations
    * binary
* Security Considerations (https://www.rfc-editor.org/rfc/rfc6838.html#page-15)
    * no provision for directives that institute actions on a recipient's files or other resources
    * no provision for directives that lead to disclosure of information
    * highlight opportunity for use of compression, highlight that the application layer: netCDF handles decompression on data access 
    * no need for services for security assurance
* Interoperability Considerations
    * no known limitations
* Published specification (https://www.rfc-editor.org/rfc/rfc6838.html#page-18 )
    * netCDF Specification reference
* Application Usage
    * scientific data sharing
    * multidimensional arrays of numerical data
* Fragment Identifier Considerations
    * no fragment identifier considerations
* Restrictions on Usage
    * no restrictions on use
* Provisional Registrations
    * unclear if this is required, worthy of consideration 
* Additional Information
    * Magic numbers
        * ?
    * File Extensions
        * .nc
* intended Use
    * common

**Comment by marqh** [comment]
@WardF, @lesserwhirls, @dopplershift, @ethanrd, @jyucsiro, @adamml 

Does this look like the making of a submission to IANA?

What could Jonathan, Adam or i do to help or facilitate this submission?

**Comment by marqh** [comment]
https://www.rfc-editor.org/rfc/rfc6838.html#page-20 describes the provisional registration process:

> A provisional registration MAY be submitted to IANA for standards-tree types.  The only required fields in such registrations are the media type name and contact information (including the standards-related organization name). 
> Upon receipt of a provisional registration, IANA will check the name and contact information, then publish the registration in a distinct publicly visible provisional registration list.

Would you be interested in expediting a provisional registration to facilitate us updating the OGC work to include this whilst we work through the formal registration details & process?

**Comment by ethanrd** [comment]
Hi @marqh - I reviewed the registration process a bit last week and expanded a bit on your application material [here](https://github.com/ethanrd/netcdf/blob/media-type-registration/MediaTypeRegistration/application-netcdf.md). I'll create a PR shortly.

I was wondering if a provisional registration would be useful. I'm open to moving forward with that if it helps the netCDF-LD effort in terms of using a media type in your OGC docs. Thoughts @WardF, @lesserwhirls, @dopplershift, @DennisHeimbigner ?

**Comment by lesserwhirls** [comment]
I would support pushing a provisional registration forward.

**Comment by ethanrd** [comment]
@marqh and all - RFC 6838 [section 4.4](https://www.rfc-editor.org/rfc/rfc6838.html#section-4.4) “Canonicalization and Format Requirements” says:

> All registered media types MUST employ a single, canonical data format, regardless of registration tree.

Thoughts on implications for netCDF-3 and netCDF-4?

I'd prefer a single media type. Most current netCDF implementations can read either so don't need a media type to tell them the difference. However, the nc-3 and nc-4 are quite different and so perhaps should be separate. (Too bad HDF5 doesn't already have a media type. Then we could consider registering "application/netcdf" and "application/netcdf+hdf5".)

**Comment by ethanrd** [comment]
Hi @marqh - I agree, a single netCDF media type seems the way to go.

I will look into requesting a provisional registration.

**Comment by ethanrd** [comment]
Hi @marqh, @ogcscott, and all,

I heard back from IANA:

> We've added the following entry to the provisional standard media type registry:
> 
> Media Type: application/netcdf
> Organization: Unidata
> Reference: Ethan Davis <edavis@ucar.edu>
> 
> Please see
> https://www.iana.org/assignments/provisional-standard-media-types

So we can start using "application/netcdf" as the netCDF media type.

Next step is to refine the draft registration information in PR #45 and get it ready to submit to IANA for the full registration. I will plan to make another pass at that in the next month or so.

Cheers,
Ethan

**Comment by marqh** [comment]
that's great progress @ethanrd 
we'll put updates in place for this 
many thanks
mark

**Comment by jerstlouis** [comment]
@ogcscotts @ethanrd Is this effort to register the official netCDF media type on-going?

It would be great if a parameter was defined to pick a particular version, since they're not backward compatible and considerably different. In the [OGC 10-090r3 Core encoding standard](https://portal.ogc.org/files/?artifact_id=43732) these are called "encoding extension" which are defined separately (e.g. in [OGC 10-092r3 NetCDF Binary Encoding Extensions Standard: NetCDF Classic and 64-bit Offset Format](https://portal.ogc.org/files/?artifact_id=43734)).
 I also don't see a reference to the netCDF 4/5 from https://www.ogc.org/standard/netcdf/ , which use [HDF5](https://www.ogc.org/standard/HDF5) if I understand correctly?

For OGC API - Coverages and OGC API - DGGS, we need to define a clear mechanism on how client and server negotiates netCDF content, and simply `Accept: application/netcdf` may not guarantee interoperability if the client is not ready to accept all possible versions aka encoding extensions.

Thanks!

**Comment by fmigneault** [comment]
Would indeed be great to have `application/netcdf` registered with IANA, along with a `version` parameter.

For the time being, I fall back to this reference: https://www.opengis.net/def/media-type/ogc/1.0/netcdf
But we are still missing a `version` specifier for their variants.

**Comment by jerstlouis** [comment]
@fmigneault Note that we defined the `netcdf3` and `netcdf4` profiles in [OGC API - DGGS](https://docs.ogc.org/is/21-038r1/21-038r1.html#_64f9565e-65d6-062f-fefb-40b30e5ca433) (alongside the corresponding DGGS-quantized profiles) for use with negotiation by profile.

**Comment by fmigneault** [comment]
Indeed, but I need a way to indicate them in *OGC API Processes* inputs/outputs. It works for single-output requested async with `?profile=netcdf4`, but it is not clear how to do so when there are multiple outputs or when they are embedded in `{"href": "...", "type": "application/netcdf"}` results. Could be a `"format": "netcdf4"`, but I would prefer referring to a registered URI.

**Comment by jerstlouis** [comment]
@fmigneault  If the server does not support after-the-fact media type negotiation of the results/outputs to request netcdf3 vs. netcdf4, perhaps we're missing something like a `"profile"` there in the execution request for [Content negotiation by profile](https://docs.ogc.org/DRAFTS/18-062r3.html#con-neg-by-profile)?

**Comment by fmigneault** [comment]
@jerstlouis  Yes. Let's open an issue OAP-side to track this separately.

**Comment by ogcscotts** [comment]
hmm, sorry that somehow I did not see this earlier! OGC can register a media type. When we do so, we point to the latest version of the Standard as published by OGC.

**Comment by jerstlouis** [comment]
@ogcscotts There are valid reasons to wish to use netcdf3 which is a lot lighter to implement due to the large number of libraries needed to support HDF5 (light web clients in particular). So a `version` parameter to the media type would be very useful.

**Comment by fmigneault** [comment]
I guess the best would be to follow the same conventions as CRSs and relation links, where `.../0/netcdf` would refer to the latest, and  `.../3/netcdf` and  `.../4/netcdf` versions would identify specific ones? Or are these numbers reserved for meta-specification updates, not the underlying spec they describe?

**Comment by jerstlouis** [comment]
@fmigneault Which URIs types are you talking about? The idea with the profile for OGC API - DGGS was that there would be a corresponding `/profile/` URI for each profile.

http://www.opengis.net/def/profile/OGC/0/netcdf3

http://www.opengis.net/def/profile/OGC/0/netcdf4

**Comment by fmigneault** [comment]
I was thinking of `http://www.opengis.net/def/profile/OGC/{3,4}/netcdf`, but the specific `netcdf{3,4}` work as well.

**Comment by jerstlouis** [comment]
@fmigneault The assumption when registering these profiles was that the Naming Authority would set up the corresponding profile definitions on the definition server at those specific URIs.  The "real" identifiers are the full URIs `https://www.opengis.net/def/profile/OGC/0/netcdf3`, `https://www.opengis.net/def/profile/OGC/0/netcdf4`, `https://www.opengis.net/def/profile/OGC/0/netcdf4-dggs`... (following the lead of [JSON-FG](https://docs.ogc.org/DRAFTS/21-045r1.html#profiles))

**Comment by ethanrd** [comment]
Hi all - Sorry this issue has been stagnant for so long. We (Unidata) [provisionally registered](https://www.iana.org/assignments/provisional-standard-media-types/provisional-standard-media-types.xhtml) `application/netcdf` with IANA back in 2022 ([comment above](https://github.com/Unidata/netcdf/issues/42#issuecomment-1007618822)). I would like to push that forward and happy to look at adding a version parameter. The version aspect gets a bit tricky as netCDF has two main data models (classic and enhanced) and several encodings: three nc-3 variants (CDF-1, CDF-2, CDF-5), nc-4, and ncZarr (fairly new, currently limited to Zarr v2).

I'm unclear on how IANA registration interacts with the OGC profiles and URIs you are discussing above.

**Comment by fmigneault** [comment]
I would propose using `profile` parameter, as in `application/netcdf; version=3; profile=cdf-1` for example.
That would follow a similar approach to `image/tiff; application=geotiff; profile=cloud-optimized`.
